python/cudf/tests/test_dataframe_copy.py

```diff
@@ -0,0 +1,127 @@
+    assert_eq(pdf['b'], copy_pdf['b'])
+    assert_eq(gdf['b'], copy_gdf['b'])
+
+
+@pytest.mark.parametrize('copy_parameters', [
+    {'fn': lambda x: x.copy(), 'expected': False},
+    {'fn': lambda x: x.copy(deep=True), 'expected': False},
+    {'fn': lambda x: copy(x), 'expected': False},
+    {'fn': lambda x: deepcopy(x), 'expected': False},
+    {'fn': lambda x: x.copy(deep=False), 'expected': True},
+    ])
+def test_dataframe_copy(copy_parameters):
+    pdf = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
+                       columns=['a', 'b', 'c'])
+    gdf = DataFrame.from_pandas(pdf)
+    copy_pdf = copy_parameters['fn'](pdf)
+    copy_gdf = copy_parameters['fn'](gdf)
+    copy_pdf['b'] = [0, 0, 0]
+    copy_gdf['b'] = [0, 0, 0]
+    pdf_pass = np.array_equal(pdf['b'].values, copy_pdf['b'].values)
+    gdf_pass = np.array_equal(gdf['b'].to_array(), copy_gdf['b'].to_array())
+    assert gdf_pass == copy_parameters['expected']
+    assert pdf_pass == copy_parameters['expected']
+
+
+"""
+DataFrame copy bounds checking - sizes 0 through 10 perform as expected
+"""
+
+
+@pytest.mark.parametrize('copy_fn', [
+    lambda x: x.copy(),
+    lambda x: x.copy(deep=True),
+    lambda x: copy(x),
+    lambda x: deepcopy(x),
+    lambda x: x.copy(deep=False),
+    ])
+@pytest.mark.parametrize('ncols', [0, 1, 2, 10])
+@pytest.mark.parametrize(
+    'data_type',
+    ['int8', 'int16', 'int32', 'int64', 'float32', 'float64',
+        'category', 'datetime64[ms]', ]
+)
+def test_cudf_dataframe_copy(copy_fn, ncols, data_type):
+    pdf = pd.DataFrame()
+    for i in range(ncols):
+        pdf[chr(i+ord('a'))] = pd.Series(np.random.randint(0, 1000, 20))\
+                                         .astype(data_type)
+    df = DataFrame.from_pandas(pdf)
+    copy_df = copy_fn(df)
+    assert_eq(df, copy_df)
+
+
+@cuda.jit
+def group_mean(data, segments, output):
+    i = cuda.grid(1)
+    if i < segments.size:
+        s = segments[i]
+        e = (segments[i + 1]
+             if (i + 1) < segments.size
+             else data.size)
+        # mean calculation
+        carry = 0.0
+        n = e - s
+        for j in range(s, e):
+            carry += data[j]
+        output[i] = carry / n
+
+
+@cuda.jit
+def add_one(data):
+    i = cuda.grid(1)
+    if i == 1:
+        data[i] = data[i] + 1.0
+
+
+def test_kernel_deep_copy():
+    pdf = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
+                       columns=['a', 'b', 'c'])
+    gdf = DataFrame.from_pandas(pdf)
+    cdf = gdf.copy(deep=True)
+    sr = gdf['b']
+    add_one[1, len(sr)](sr.to_gpu_array())
+    assert not gdf.to_string().split() == cdf.to_string().split()
+
+
+def test_kernel_shallow_copy():
+    pdf = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
+                       columns=['a', 'b', 'c'])
+    gdf = DataFrame.from_pandas(pdf)
+    cdf = gdf.copy(deep=False)
+    sr = gdf['a']
+    add_one[1, len(sr)](sr.to_gpu_array())
+    assert_eq(gdf, cdf)
```

Comment:
After this I'd add a test that assigning a new column to cdf doesn't create the column in gdf.

Reply:
See test_cudf_dataframe_copy_then_insert.